### ISSUE TITLE
Parametrization goes into infinite recursion when trying to print a module

### ISSUE THREAD

**Issue by z-a-f** [pytorch/pytorch#59926]
## 🐛 Bug
One use case for parametrization is to use some buffers within existing layers:

```python
class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(x)

model = Model()
model.linear.register_buffer('mask', torch.ones(model.linear.weight.shape))

class WeightMaskParametrization(nn.Module):
    def __init__(self, layer):
        super().__init__()
        self.layer = layer
        
    def forward(self, w):
        return self.layer.mask * w

weight_mask = WeightMaskParametrization(model.linear)
parametrize.register_parametrization(model.linear, 'weight', weight_mask);  # Notise the semicolon here

print(model)  # <- this will throw a recursion error
```

The error is in the `nn/modules/module.py:1741`, which recursively calls the print statement on the children. I think because the parametrized layer tries to print its internals, it recursively calls the repr on the parametrization.

---

One solution is to put the tensor as the argument to the input:

```python

class WeightMaskParametrization(nn.Module):
    def __init__(self, mask):
        super().__init__()
        self.mask = mask
        
    def forward(self, w):
        return self.mask * w

weight_mask = WeightMaskParametrization(model.linear.mask)
parametrize.register_parametrization(model.linear, 'weight', weight_mask)
```

The problem with that approach is if the tensor is reassigned, the parametrization will lose track of it.

```python
weight_mask = WeightMaskParametrization(model.linear.mask)
parametrize.register_parametrization(model.linear, 'weight', weight_mask)
model.linear.mask = torch.zeros(model.linear.weight.shape)

model.linear.weight  # <- This will use the old mask tensor
```

cc @albanD @mruberry @jbschlosser

**Comment by albanD** [comment]
cc @Lezcano

**Comment by albanD** [comment]
Note that this is not specific to parametrization, anything your do such a cycle of Modules, you will get a recursion error when printing:
```python
import torch

a = torch.nn.Linear(2, 2)
b = torch.nn.Linear(2, 2)

a.b = b
b.a =a 

print(a)  # RecursionError
```

**Comment by Lezcano** [comment]
For reference, I believe that the second proposed approach is the correct one. One can make it work modifying just the internal tensor storage via the function `set_` (https://pytorch.org/docs/stable/generated/torch.Tensor.set_.html). This should work in the case when, for example, you are sharing the mask between a number of layers. You would then set a new mask as:
```python
model.linear.mask.set_(torch.zeros(model.linear.weight.shape))
```

A simplified implementation of your example could then be:
```python
import torch
import torch.nn as nn
import torch.nn.utils.parametrize as parametrize

class WeightMaskParametrization(nn.Module):
    def __init__(self, mask):
        super().__init__()
        self.mask = mask

    def forward(self, w):
        return self.mask * w

model = nn.Linear(3, 4)
model.register_buffer('mask', torch.ones_like(model.weight))
parametrize.register_parametrization(model, 'weight', WeightMaskParametrization(model.mask))

print(model.weight)   # print original weight
model.mask.set_(torch.zeros_like(model.weight))
print(model.weight)   # print zeros
```

In this simplified example, it would be even better to have the `mask` live in the class `WeightMaskParametrization` rather than in `nn.Linear`, and thus, be managed directly by this class, but I guess that the real example motivating this issue may not allow for that.

Another thing to point out is that this `WeightMaskParametrization` is just pruning in disguise. We will eventually move all the pruning methods into parametrizations, and the base parametrization will certainly take the form of `WeightMaskParametrization`, with a few more bells and whistles.

**Comment by Lezcano** [comment]
I'm closing this for now, but do feel free to reopen.
